### PR TITLE
Prevent duplicate contention migration (fix #2054)

### LINKED ISSUE #2054: Migrate contention data from the EP400 to the pending EP
Once the anchors are set on the EP400 claim to prevent it from being assigned to a VSR while we process it, migrate the contention and  contention data from the EP400 to the pending EP.

## **Acceptance Criteria**
Within **EP Merge App (VRO)**
- Check whether the contention on the EP400 is a duplicate of the contentions on the pending EP. A contention is considered duplicate if:
  - The contention name is exactly the same, and
  - The contention type is the same (increase, new, etc.)
- If the contention **is not a duplicate**:
  - Migrate the contention and contention data from the EP400 to the pending EP (the 010, 110, or 020)
  - Contention data = contention classification, contention date, and all special issues added to that contention
- If the contention **is a duplicate**:
  - Do not migrate the contention and proceed to #2056 

See #2057 for error handling requirements.

### PR DESCRIPTION
**Pull request by raudabaugh**
## What was the problem?

The merge app did not check if an EP400 contention already existed in the pending claim before migrating it.

Associated tickets or Slack threads:
- #2054 

## How does this fix it?[^1]

The merge job skips migrating a contention if there is already one in the pending claim with the same `claimantText` and `contentionTypeCode`.

## How to test this PR
- From the `abd-vro` directory, run the unit tests:
`./gradlew :domain-ee:test`

[^1]: [Pull-Requests guidelines](https://github.com/department-of-veterans-affairs/abd-vro/wiki/Pull-Requests). If PR is significant, update [Current Software State](https://github.com/department-of-veterans-affairs/abd-vro/wiki/Current-Software-State) wiki page.
[^secrel]: To check if a PR will succeed in the SecRel workflow, [test PRs in the SecRel pipeline](https://github.com/department-of-veterans-affairs/abd-vro-internal/wiki/Secure-Release-process#to-test-prs-in-the-secrel-pipeline).

### CODE REVIEW

File: domain-ee/ee-ep-merge-app/src/python_src/service/ep_merge_machine.py [hunk 1]
```diff
@@ -78,6 +79,8 @@ def on_get_ep400_contentions(self, pending_contentions=None):
 
     @running_set_temp_station_of_jurisdiction.enter
     def on_set_temp_station_of_jurisdiction(self, pending_contentions=None, ep400_contentions=None):
```
Comment (dfitchett) [inline domain-ee/ee-ep-merge-app/src/python_src/service/ep_merge_machine.py:81]: So on line 89 `pending_contentions` and `ep400_contentions` are both added to the self.process(...) function as kwargs. So they are actually available for you to use in the function signature of is_duplicate as kwargs: `is_duplicate(self, pending_contentions, ep400_contentions)`.  So you dont actually need to add them to the MergeJob class. 🪄 🧙

File: domain-ee/ee-ep-merge-app/src/python_src/model/merge_job.py [hunk 2]
```diff
@@ -23,6 +25,8 @@ class MergeJob(BaseModel):
     state: JobState = JobState.PENDING
     error_state: JobState | None = None
     messages: list[Any] | None = None
+    pending_contentions: list[ContentionSummary] | None = None
```
Comment (dfitchett) [inline domain-ee/ee-ep-merge-app/src/python_src/model/merge_job.py:28]: These additions are unnecessary. See comment on ep_merge_machine.py 👀

File: domain-ee/ee-ep-merge-app/src/python_src/service/ep_merge_machine.py [hunk 3]
```diff
@@ -36,7 +36,8 @@ class EpMergeMachine(StateMachine):
             | running_get_pending_contentions.to(completed_error, cond="has_error")
             | running_get_ep400_contentions.to(running_set_temp_station_of_jurisdiction, unless="has_error")
             | running_get_ep400_contentions.to(completed_error, cond="has_error")
-            | running_set_temp_station_of_jurisdiction.to(running_merge_contentions, unless="has_error")
+            | running_set_temp_station_of_jurisdiction.to(running_merge_contentions, unless=["is_duplicate", "has_error"])
```
Comment (dfitchett) [inline domain-ee/ee-ep-merge-app/src/python_src/service/ep_merge_machine.py:39]: Nice use of these guards 👍

File: domain-ee/ee-ep-merge-app/src/python_src/service/ep_merge_machine.py [hunk 4]
```diff
@@ -143,6 +146,9 @@ def make_request(self,
     def has_error(self):
         return self.job.state == JobState.COMPLETED_ERROR
 
+    def is_duplicate(self):
+        return not ContentionsUtil.new_contentions(self.job.pending_contentions, self.job.ep400_contentions)
```
Comment (dfitchett) [inline domain-ee/ee-ep-merge-app/src/python_src/service/ep_merge_machine.py:150]: Should we catch the CompareException here and call the log_error function?

File: domain-ee/ee-ep-merge-app/src/python_src/util/contentions_util.py [hunk 5]
```diff
@@ -7,9 +8,22 @@ def __init__(self, message=None):
...
+        self.message = message
+        super().__init__(self.message)
+
+
 class ContentionsUtil:
     @staticmethod
     def merge_claims(pending_contentions: Response = None, ep400_contentions: Response = None):
         if pending_contentions and ep400_contentions:
-            return [*pending_contentions.contentions, *ep400_contentions.contentions]
+            return [*pending_contentions.contentions, *ContentionsUtil.new_contentions(pending_contentions.contentions, ep400_contentions.contentions)]
```
Comment (dfitchett) [inline domain-ee/ee-ep-merge-app/src/python_src/util/contentions_util.py:21]: If for some reason this method is used before any call to new_contentions, there's a possiblity of the `new_contentions` raising CompareException. Should it be caught here?

File: domain-ee/ee-ep-merge-app/integration/test_merge_request.py
Comment (dfitchett) [inline domain-ee/ee-ep-merge-app/integration/test_merge_request.py:1]: Can we get some integration tests added here for testing the duplicate checking?

**Review by dfitchett — CHANGES_REQUESTED** [review]

File: domain-ee/ee-ep-merge-app/src/python_src/service/ep_merge_machine.py (hunk 1 above)
Comment (raudabaugh) [reply domain-ee/ee-ep-merge-app/src/python_src/service/ep_merge_machine.py:81]: I was thinking those might be useful to have for restarts/recovery when we get to converting the job store to RDS or some persistent backend. But we could definitely leave those out until we get to that point - thoughts?

Comment (dfitchett) [reply domain-ee/ee-ep-merge-app/src/python_src/service/ep_merge_machine.py:81]: Noted! I'd say for now, lets just leave them out! We can have another conversation when it comes to the data we should/can/can't store in RDS!

**Review by dfitchett — APPROVED** [review]
Very Nice! 👍 LGTM!

File: domain-ee/ee-ep-merge-app/src/python_src/util/contentions_util.py [hunk 6]
```diff
@@ -7,9 +8,25 @@ def __init__(self, message=None):
...
+
 class ContentionsUtil:
     @staticmethod
     def merge_claims(pending_contentions: Response = None, ep400_contentions: Response = None):
         if pending_contentions and ep400_contentions:
-            return [*pending_contentions.contentions, *ep400_contentions.contentions]
+            try:
+                return [*pending_contentions.contentions, *ContentionsUtil.new_contentions(pending_contentions.contentions, ep400_contentions.contentions)]
+            except CompareException:
+                raise
```
Comment (nanotone) [inline domain-ee/ee-ep-merge-app/src/python_src/util/contentions_util.py:24]: Is there any reason to catch this `CompareException` and re-raise the original? It seems easier to just leave out the try, except, and raise lines.

**Review by nanotone — APPROVED** [review]
Approving, with two non-blocking comments:
- The one about re-raising an exception
- Could you go through and correct the `tinitus` typo to `tinnitus`, in code and in filenames?

**Review by dfitchett — APPROVED** [review]

**Pull request merged**